lib/id3c/api/datastore.py

```diff
@@ -256,38 +256,55 @@ def fetch_identifier_set(session: DatabaseSession, name: str) -> Any:
 
 @export
 @catch_permission_denied
-def make_identifier_set(session: DatabaseSession, name: str, **fields) -> bool:
+def make_identifier_set(session: DatabaseSession, name: str, use: str, **fields) -> bool:
     """
-    Create a new identifier set *name* in the backing database using *session*
-    if it doesn't already exist.
+    Create a new identifier set *name* with *use* in the backing database
+    using *session* if it doesn't already exist.
 
     If *description* is provided as a keyword argument, its value is
-    set/updated in the database.
+    set/updated in the database. If identifier set with *name* already exists,
+    the *description* is only updated if *use* matches the existing value.
 
     Returns ``True`` if the set was created or updated and ``False`` if it
-    already existed.
+    already existed or was not updated due to mismatched use.
     """
     with session, session.cursor() as cursor:
         # If I expected to have additional columns in the future, I'd build up
         # the SQL dynamically, but I don't, so a simple conditional is enough.
         #   -trs, 1 July 2021
         if "description" in fields:
             cursor.execute("""
-                insert into warehouse.identifier_set (name, description)
-                    values (%s, nullif(%s, ''))
+                insert into warehouse.identifier_set (name, use, description)
+                    values (%s, %s, nullif(%s, ''))
                     on conflict (name) do update
                         set description = excluded.description
-                """, (name, fields["description"]))
+                        where identifier_set.use = excluded.use
+                """, (name, use, fields["description"]))
 
         else:
             cursor.execute("""
-                insert into warehouse.identifier_set (name)
-                    values (%s)
+                insert into warehouse.identifier_set (name, use)
+                    values (%s, %s)
                     on conflict (name) do nothing
-                """, (name,))
+                """, (name, use))
```

Comment:
I'm on the fence about whether its best to make `use` a set-once attribute or leave it up to usage/convention. I see the upside of not letting an identifier set change uses (that's potentially confusing if the identifiers have already been used for that purpose!), but outside of this API, there's no restriction on updating it, so it's a little weird to enforce it here but not elsewhere. My thinking is leaning towards not requiring it to be set-once.

If we _do_ require it to be set-once, then I think a 409 Conflict response should be returned when the value passed in doesn't match the database (during both creation and update). Otherwise a client doesn't get useful information about the state of the resource and has to conditionalize on a fetch of the identifier set instead of just making one idempotent PUT request. I also think that if we require it set-once in the API, we should strongly consider propagating that restriction deeper into the database itself.

Reply:
That makes sense. I didn't consider that the API is already locked down to the equivalent database permissions, so currently only postgres user or role members will be able to use this endpoint. So I agree we can allow updates to `use` here.